### PR TITLE
add quarantine check to scan for quarantine files and meddlesome processes

### PR DESCRIPTION
**Pull request by James-Pickett**
This check reports and files that have "quarantine" in their path. It recursively checks each of the provided root dirs to the specified depth.

example summary:
❌	Quarantine: found 1 quarantined files

example log:
```
starting quarantine check

possilby meddlesome proccesses:
map[cmdline:[] create_time:1693926941054 exe:C:\ProgramData\Microsoft\Windows Defender\Platform\4.18.23080.2006-0\NisSrv.exe pid:9268 ppid:1236 status:error: not implemented yet]
map[cmdline:[] create_time:1693926936001 exe:C:\ProgramData\Microsoft\Windows Defender\Platform\4.18.23080.2006-0\MsMpEng.exe pid:15636 ppid:1236 status:error: not implemented yet]

searching for quarantined files:
failed to read C:\ProgramData\Microsoft\Windows\CapabilityAccessManager: open C:\ProgramData\Microsoft\Windows\CapabilityAccessManager: Access is denied.
/Library/Application Support does not exist
total dirs checked: 1159
quarantine directory paths and files:
C:\ProgramData\Microsoft\Windows Defender\Quarantine: 0 files
C:\ProgramData\quarantine: 1 files
  test.txt

```

### CODE REVIEW

File: pkg/debug/checkups/quarantine.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+			q.checkDirs(extraFh, currentDepth+1, maxDepth, filepath.Join(dirPath, dirEntry.Name()), directoryKeyword)
+			continue
+		}
+
+		if !dirNameContainsKeyword {
+			// not in quarantine dir
+			continue
+		}
+
+		q.quarantineCounts[dirPath]++
```
Comment (RebeccaMahany) [inline pkg/debug/checkups/quarantine.go:135]: Do we want to list the names of the quarantined files? It looks like we're only doing count per directory, but maybe I'm missing it?

File: pkg/debug/checkups/quarantine.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		totalQuarantinedFiles += count
+	}
+
+	if totalQuarantinedFiles == 0 {
+		q.status = Passing
+		q.summary = "no files found in quarantine directories"
+		return nil
+	}
+
+	q.status = Failing
```
Comment (RebeccaMahany) [inline pkg/debug/checkups/quarantine.go:91]: What about looking specifically for launcher/launcher.exe and osqueryd/osqueryd.exe being quarantined, and only setting the status to `Failing` if one of those is quarantined? I don't know if some other non-Kolide-related file being quarantined should result in a `Failing` status.

File: pkg/debug/checkups/quarantine.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+type quarantine struct {
+	status           Status
+	summary          string
+	quarantineCounts map[string]int
+	dirsChecked      int
+}
+
+func (q *quarantine) Name() string {
+	return "Quarantine"
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go:35]: I added a linux path.

File: pkg/debug/checkups/quarantine.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func (q *quarantine) Run(ctx context.Context, extraFh io.Writer) error {
+	q.quarantineCounts = make(map[string]int)
+
+	var (
+		quarantinePathDepths = map[string]int{
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go]: done

File: pkg/debug/checkups/quarantine.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+			// not in quarantine dir
+			continue
+		}
+
+		q.quarantineCounts[dirPath]++
+	}
+}
+
+func (q *quarantine) logMeddlesomeProccesses(ctx context.Context, extraFh io.Writer, containsSubStrings []string) error {
+	fmt.Fprint(extraFh, "\npossilby meddlesome proccesses:\n")
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go]: fixed

File: pkg/debug/checkups/quarantine.go (hunk 1 above)
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go:168]: I started off with that but learned that AVs tend to renamed the files to GUIDs then keep some metadata some where. I don't know if they all do this, only learned from seph that crowdstrike does this. Windows defender does the same.

File: pkg/debug/checkups/quarantine.go (hunk 2 above)
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go:122]: Same as above, AVs renamed the files to GUIDs and store metadata some where so can't search for osquery directly. I actually started off on this path. However, it raises and interesting question. Should we ever fail this check or should we keep at as information purposes only?

File: pkg/debug/checkups/quarantine.go (hunk 1 above)
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go:168]: After further discussion, re adding file names, even if their just guid's it wont hurt.

File: pkg/debug/checkups/quarantine.go (hunk 2 above)
Comment (directionless) [reply pkg/debug/checkups/quarantine.go:126]:
```suggestion
	q.status = Warning
```

I'm not sure we know enough to declare this as a failing. It's a warning or informational

File: pkg/debug/checkups/quarantine.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		meddlesomeProcessPatterns = []string{
+			`crowdstrike`,
+			`opswat`,
+			`defend`,
+			`defense`,
+			`threat`,
+			`virus`,
+			`quarantine`,
+			`snitch`,
+		}
```
Comment (directionless) [inline pkg/debug/checkups/quarantine.go:76]: Do we know what carbonblack's process name is?

File: pkg/debug/checkups/quarantine.go (hunk 2 above)
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go]: updated to warning

File: pkg/debug/checkups/quarantine.go (hunk 6 above)
Comment (James-Pickett) [reply pkg/debug/checkups/quarantine.go:84]: googling revealed many possibilities, I added all the ones that looked promising

File: pkg/debug/checkups/quarantine.go (hunk 1 above)
Comment (directionless) [reply pkg/debug/checkups/quarantine.go]: I'm fine with this, but I'll note that the test host I found had ~600 of them

**Review by directionless — APPROVED** [review]
Sure, let's try

**Pull request merged**